File: api/Enkl.Api/Services/ReleaseService.cs
using System.Text.RegularExpressions;
using Enkl.Api.Data;
using Enkl.Api.Domain.Entities;
using Enkl.Api.Dtos;
using Microsoft.EntityFrameworkCore;

namespace Enkl.Api.Services;

public class ReleaseService
{
    private readonly AppDbContext _db;

    public ReleaseService(AppDbContext db)
    {
        _db = db;
    }

    public async Task<ReleaseDto?> CreateAsync(Guid projectId, CreateReleaseRequest request)
    {
        var projectExists = await _db.Projects.AnyAsync(p => p.Id == projectId);
        if (!projectExists) return null;

        var release = new Release
        {
            Id = Guid.NewGuid(), ProjectId = projectId, Name = request.Name,
            Status = request.Status is "pending" or "in_progress" or "deployed" ? request.Status : "pending",
            OwnerId = request.OwnerId, StartDate = request.StartDate, EndDate = request.EndDate,
            Color = ResolveColor(request.Color),
            DateCreated = DateTime.UtcNow, DateLastModified = DateTime.UtcNow
        };
        _db.Releases.Add(release);
        await _db.SaveChangesAsync();
        return ToDto(release);
    }

    public async Task<ReleaseDto?> UpdateAsync(Guid projectId, Guid releaseId, UpdateReleaseRequest request)
    {
        var release = await _db.Releases.FirstOrDefaultAsync(r => r.Id == releaseId && r.ProjectId == projectId);
        if (release is null) return null;

        release.Name = request.Name;
        release.Status = request.Status is "pending" or "in_progress" or "deployed" ? request.Status : "pending";
        release.OwnerId = request.OwnerId;
        release.StartDate = request.StartDate;
        release.EndDate = request.EndDate;
        release.Color = ResolveColor(request.Color);
        release.DateLastModified = DateTime.UtcNow;
        await _db.SaveChangesAsync();
        return ToDto(release);
    }

    /// <summary>The ONLY write path for ReleaseNotes — gated by ReleasesController's own
    /// [Authorize(Policy = "ProjectAdmin")] on this one action, never via CreateAsync/UpdateAsync
    /// above (see root CLAUDE.md §7's "one-endpoint-owns-the-field" convention).</summary>
    public async Task<ReleaseDto?> UpdateNotesAsync(Guid projectId, Guid releaseId, UpdateReleaseNotesRequest request)
    {
        var release = await _db.Releases.FirstOrDefaultAsync(r => r.Id == releaseId && r.ProjectId == projectId);
        if (release is null) return null;

        release.ReleaseNotes = request.ReleaseNotes;
        release.DateLastModified = DateTime.UtcNow;
        await _db.SaveChangesAsync();
        return ToDto(release);
    }

    public async Task<bool> DeleteAsync(Guid projectId, Guid releaseId)
    {
        var release = await _db.Releases.FirstOrDefaultAsync(r => r.Id == releaseId && r.ProjectId == projectId);
        if (release is null) return false;

        // Mirrors mutations.js's deleteRelease: unassign referencing tasks rather than blocking
        // the delete (Task.ReleaseId is already configured ON DELETE SET NULL, so removing the
        // row is enough — no explicit unassign loop needed).
        _db.Releases.Remove(release);
        await _db.SaveChangesAsync();
        return true;
    }

    private static readonly Regex HexColorPattern = new(@"^#[0-9a-fA-F]{3}([0-9a-fA-F]{3})?$", RegexOptions.Compiled);

    /// <summary>Falls back to the default grey for anything not a plain #rgb/#rrggbb hex string —
    /// same "invalid input collapses to the safe default" convention as every other defensively-
    /// parsed field in this app (root CLAUDE.md §1).</summary>
    private static string ResolveColor(string? color) =>
        !string.IsNullOrWhiteSpace(color) && HexColorPattern.IsMatch(color) ? color : "#cccccc";

    private static ReleaseDto ToDto(Release r) => new(r.Id, r.Name, r.Status, r.OwnerId, r.StartDate, r.EndDate, r.ReleaseNotes, r.Color);
}

File: php-api/src/Services/ReleaseService.php
<?php

declare(strict_types=1);

namespace Enkl\Api\Services;

use Enkl\Api\Support\Uuid;
use PDO;

/** Ported from Services/ReleaseService.cs. */
final class ReleaseService
{
    private const VALID_STATUSES = ['pending', 'in_progress', 'deployed'];

    public function __construct(private readonly PDO $db)
    {
    }

    public function create(string $projectId, array $request): ?array
    {
        $stmt = $this->db->prepare('SELECT 1 FROM "Projects" WHERE "Id" = :id');
        $stmt->execute(['id' => $projectId]);
        if ($stmt->fetch() === false) {
            return null;
        }

        $id = Uuid::v4();
        $status = in_array($request['status'] ?? null, self::VALID_STATUSES, true) ? $request['status'] : 'pending';
        $color = self::resolveColor($request['color'] ?? null);
        $stmt = $this->db->prepare(<<<SQL
            INSERT INTO "Releases" ("Id", "ProjectId", "Name", "Status", "OwnerId", "StartDate", "EndDate", "Color", "DateCreated", "DateLastModified")
            VALUES (:id, :pid, :name, :status, :ownerId, :start, :end, :color, now(), now())
        SQL);
        $stmt->execute([
            'id' => $id, 'pid' => $projectId, 'name' => $request['name'] ?? '', 'status' => $status,
            'ownerId' => $request['ownerId'] ?? null, 'start' => $request['startDate'] ?? null, 'end' => $request['endDate'] ?? null,
            'color' => $color,
        ]);

        return ['id' => $id, 'name' => $request['name'] ?? '', 'status' => $status, 'ownerId' => $request['ownerId'] ?? null, 'startDate' => $request['startDate'] ?? null, 'endDate' => $request['endDate'] ?? null, 'releaseNotes' => null, 'color' => $color];
    }

    public function update(string $projectId, string $releaseId, array $request): ?array
    {
        $stmt = $this->db->prepare('SELECT "ReleaseNotes" FROM "Releases" WHERE "Id" = :id AND "ProjectId" = :pid');
        $stmt->execute(['id' => $releaseId, 'pid' => $projectId]);
        $existing = $stmt->fetch();
        if ($existing === false) {
            return null;
        }

        $status = in_array($request['status'] ?? null, self::VALID_STATUSES, true) ? $request['status'] : 'pending';
        $color = self::resolveColor($request['color'] ?? null);
        $stmt = $this->db->prepare(<<<SQL
            UPDATE "Releases" SET "Name" = :name, "Status" = :status, "OwnerId" = :ownerId,
                "StartDate" = :start, "EndDate" = :end, "Color" = :color, "DateLastModified" = now()
            WHERE "Id" = :id
        SQL);
        $stmt->execute([
            'name' => $request['name'] ?? '', 'status' => $status, 'ownerId' => $request['ownerId'] ?? null,
            'start' => $request['startDate'] ?? null, 'end' => $request['endDate'] ?? null, 'color' => $color, 'id' => $releaseId,
        ]);

        return ['id' => $releaseId, 'name' => $request['name'] ?? '', 'status' => $status, 'ownerId' => $request['ownerId'] ?? null, 'startDate' => $request['startDate'] ?? null, 'endDate' => $request['endDate'] ?? null, 'releaseNotes' => $existing['ReleaseNotes'], 'color' => $color];
    }

    /** The ONLY write path for ReleaseNotes — gated by the controller's own ProjectAdminMiddleware
     * sub-group, never via create()/update() above (see root CLAUDE.md §7's
     * "one-endpoint-owns-the-field" convention). */
    public function updateNotes(string $projectId, string $releaseId, ?string $releaseNotes): ?array
    {
        $stmt = $this->db->prepare('SELECT "Id", "Name", "Status", "OwnerId", "StartDate", "EndDate", "Color" FROM "Releases" WHERE "Id" = :id AND "ProjectId" = :pid');
        $stmt->execute(['id' => $releaseId, 'pid' => $projectId]);
        $row = $stmt->fetch();
        if ($row === false) {
            return null;
        }

        $update = $this->db->prepare('UPDATE "Releases" SET "ReleaseNotes" = :notes, "DateLastModified" = now() WHERE "Id" = :id');
        $update->execute(['notes' => $releaseNotes, 'id' => $releaseId]);

        return [
            'id' => $row['Id'], 'name' => $row['Name'], 'status' => $row['Status'], 'ownerId' => $row['OwnerId'],
            'startDate' => $row['StartDate'], 'endDate' => $row['EndDate'], 'releaseNotes' => $releaseNotes,
            'color' => $row['Color'],
        ];
    }

    public function delete(string $projectId, string $releaseId): bool
    {
        // Tasks.ReleaseId is already SetNull, so removing the row is enough — no explicit unassign
        // loop needed, mirrors mutations.js's deleteRelease.
        $stmt = $this->db->prepare('DELETE FROM "Releases" WHERE "Id" = :id AND "ProjectId" = :pid');
        $stmt->execute(['id' => $releaseId, 'pid' => $projectId]);
        return $stmt->rowCount() > 0;
    }

    /** Falls back to the default grey for anything not a plain #rgb/#rrggbb hex string — same
     * "invalid input collapses to the safe default" convention as every other defensively-parsed
     * field in this app (root CLAUDE.md §1). */
    private static function resolveColor(?string $color): string
    {
        return $color !== null && preg_match('/^#[0-9a-fA-F]{3}([0-9a-fA-F]{3})?$/', $color) === 1 ? $color : '#cccccc';
    }
}
